R/topTaxa.R

```diff
@@ -0,0 +1,90 @@
+#' Returns top N \dQuote{FeatureID}
+#'
+#' Basic extractor function for extracting top N \dQuote{FeatureID}
+#' from assay for a user specified \dQuote{SampleID} in a
+#' \code{\link[=SummarizedExperiment-class]{SummarizedExperiment}}
+#' object.
+#'
+#' @param x A
+#'  \code{\link[=SummarizedExperiment-class]{SummarizedExperiment}} object.
+#'
+#' @param top Numeric value, how many top taxa to return. Default return top five taxa.
+#'
+#' @param method Specify the method to determine top taxa. Either sum, mean or median.
+#'               Default is 'mean'
+#'
+#' @param abund_values a \code{character} value to select an
+#'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{assayNames}}
+#'
+#' @importFrom DelayedMatrixStats rowSums2 rowMeans2 rowMedians
+#'
+#' @return A vector of the top N \dQuote{FeatureID}
+#' @name topTaxa
+#'
+#' @author
+#' Sudarshan A. Shetty
+#'
+#' @examples
+#' data(GlobalPatterns, package = "MicrobiomeExperiment")
+#'
+#' top_taxa <- topTaxa(GlobalPatterns,
+#'                     method="mean",
+#'                     top=5,
+#'                     abund_values="counts")
+#'
+NULL
+
+#' @rdname topTaxa
+#'
+#' @export
+
+setGeneric("topTaxa", signature = "x",
```

Comment:
The name of this function does not follow convention. It does not contain a verb, which is required for a function not accessing a slot of the main input object. More applicable would be `getTopTaxa`.

Please also note that this is basically the same functionality as `getTopHVGs` from the `scran` package. It would make much more sense, to write a wrapper around `getTopHVGs`, since that is a tried and tested function.

The additional benefit you could provide is a reference to the taxonomic rank. For example you could implement the base wrapper around `getTopHVGs` for `SummarizedExperiment` and implement the function for `MicrobiomeExperiment` as well. In the latter function you could then write something like this:

```
top_taxa <- callNextMethod() # This defers to the normal SummarizedExperiment method
# select top_taxa based on a selected rank
# e.g. translated the returned row names to taxonomic values of a selected rank.
# This way you answer the question of "What are the top Phyla for the most 10 abundant ASV?"
```

Reply:
The `getTopHVGs` functions has a different purpose than `getTopTaxa`. Since the input for variable taxa may have multiple options in microbiome like the output from `DESeq2`, `ALDEx2`, `ANCOM` etc. it may be better to have a separate function that includes `getTopHVGs` for a generally applicable utility. `getTopHVGs` ar similar will be useful for instance when plotting dendrogram-heatmaps, fold-change plots. While `getTopTaxa` is useful for barplots, boxplots, and summarising overall data. 
Do you think the taxonomy option can be more general like `getTaxonomy` that will take input as the output of `getTopTaxa`  or any user-specified list and return classification at a specific or all taxonomic levels? This can come in handy when wanting to analyze/visualize at ASVs level. I currently do use [`format_to_besthit`()](https://microsud.github.io/microbiomeutilities/articles/microbiomeutilities.html#formatting-the-phyloseq-object) when working with ASV/OTU level data. Something we can think about and which in my experience are common questions "Which are the `n` most dominant `ASVs` in each of the phyla/family/...?" and "Which are the n most dominant taxa in each of the samples?". Let me know your thoughts.

Reply:
Ok my bad. `getTopHVGs` returns the row names of the most variable rows and uses recalculate statistics in a `DataFrame`, whereas `getTopTaxa` focuses on the overall abundance in assay data.

Underlying this mixup is a different definition of the word `top`. top in most interesting because of most changes or top as in must abundant. Am I right?

Regarding the taxonomic option: Please have a look at the implementation of `agglomerateByRank`. There the `rank` argument directs to the taxonomic level to be used. Translated to `getTopTaxa` this could result in the following:

```
setMethod("getTopTaxa", signature = c(x = "MicrobiomeExperiment"),
    function(x, rank = NULL, ...){
     ans <- callNextMethod(...)
     if(!is.null(rank)){
         if(length(rank) != 1L || !(rank %in% taxonomyRanks(x))){
            stop("If set, 'rank' must be a valid taxonomic rank as returned by taxonomyRanks(x)", call. = FALSE)
         }
         tax_data <- rowData(x)[,taxonomyRanks(x)]
         ans <- unique(tax_data[ans])
     }
     ans
  }
)
```

So if `rank` is not set, it returns a row name of the `SE` object and if it is set a look in the `rowData` starts to return the top taxonomic rank values for the given rank. Would that make sense to you?

As a side note since you brought it up:
Can you explain to me the function of of `format_to_besthit()`? As far as I can tell from the manual page it reformats the `phyloseq` row names, to lowest populated taxonomic level? If that is the case, this function is available as `getTaxonomyLabel` in `mia`